File: benchmarks/python/macbook/bounce.py
TAG = "bounce"
WARMUP = 1
CYCLE = 4

RESULT = 1331

# random
class Random:
    def __init__(self):
        self._seed = 74755

    def next(self):
        self._seed = ((self._seed * 1309) + 13849) & 65535

        return self._seed


class Ball:
    def __init__(self, random):
        self._x = random.next() % 500
        self._y = random.next() % 500
        self._x_vel = (random.next() % 300) - 150
        self._y_vel = (random.next() % 300) - 150

    def bounce(self):
        x_limit = 500
        y_limit = 500
        bounced = False

        self._x += self._x_vel
        self._y += self._y_vel

        if self._x > x_limit:
            self._x = x_limit
            self._x_vel = -abs(self._x_vel)
            bounced = True

        if self._x < 0:
            self._x = 0
            self._x_vel = abs(self._x_vel)
            bounced = True

        if self._y > y_limit:
            self._y = y_limit
            self._y_vel = -abs(self._y_vel)
            bounced = True

        if self._y < 0:
            self._y = 0
            self._y_vel = abs(self._y_vel)
            bounced = True

        return bounced

def bounce():
    random = Random()

    ball_count = 100
    bounces = 0
    balls = [None] * ball_count

    for i in range(ball_count):
        balls[i] = Ball(random)

    for i in range(50):
        for ball in balls:
            if ball.bounce():
                bounces += 1
    return bounces            

def verify_result(result):
    return result == RESULT

def benchmark_main():
    result = bounce()
    verify_result(result)

File: benchmarks/c/esp32/main/bounce.c
#include <stdint.h>
#include <stdlib.h>
#include <stdbool.h>
#include <assert.h>
#include <stdio.h>

#define WARMUP 0
#define CYCLE 3
#define TAG "bounce"


#define RESULT 1331
#define BALL_COUNT 100


// random
int random_seed = 74755;

int random_next() {
    random_seed = ((random_seed * 1309) + 13849) & 65535;
    return random_seed;
}

// finish random

typedef struct ball {
    int x;
    int y;
    int x_vel;
    int y_vel;
} ball_t;

bool ball_bounce(ball_t *ball) {
    int x_limit = 500;
    int y_limit = 500;
    bool bounced = false;

    ball->x += ball->x_vel;
    ball->y += ball->y_vel;

    if (ball->x > x_limit) {
        ball->x = x_limit;
        ball->x_vel = -abs(ball->x_vel);
        bounced = true;
    }
    if (ball->x < 0) {
        ball->x = 0;
        ball->x_vel = abs(ball->x_vel);
        bounced = true;
    }
    if (ball->y > y_limit) {
        ball->y = y_limit;
        ball->y_vel = -abs(ball->y_vel);
        bounced = true;
    }
    if (ball->y < 0) {
        ball->y = 0;
        ball->y_vel = abs(ball->y_vel);
        bounced = true;
    }
    return bounced;
}

int bounce() {
    random_seed = 74755;
    int ball_count = BALL_COUNT;
    int bounces = 0;
    // ball_t balls[ball_count];
    ball_t *balls = malloc(sizeof(ball_t) * ball_count);
    for (int i = 0; i < ball_count; i++) {
        ball_t ball = {
            random_next() % 500,
            random_next() % 500,
            (random_next() % 300) - 150,
            (random_next() % 300) - 150
        };
        balls[i] = ball;
    }
    for (int i = 0; i < 50; i++) {
        for (int b = 0; b < ball_count; b++) {
            if (ball_bounce(&balls[b])) {
                bounces += 1;
            }
        }
    }
    free(balls);
    return bounces;
}

bool verify_result(int result) {
    return result == RESULT;
}

void benchmark_main() {
    int result = bounce();
    assert(verify_result(result));
}
